issue 4179: sort output of restic ls -l by size, atime, ctime, mtime, time(=mtime), extension

wplapper opened the pull request:
Enhancement: create ability to sort output of restic ls -l by name, size, atime, ctime, mtime, time(=mtime), X(=extension), extension

https://github.com/restic/restic/issues/4179

<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
Ability to sort output of restic ls -l by name, size, atime, ctime, mtime, time(=mtime)
<!--
Describe the changes and their purpose here, as detailed as needed.
-->
Add 2 local parameters to sort command: --sort and --reverse, so that the output can be sorted by one of the above named possibilities.
Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->
https://github.com/restic/restic/issues/4179
closes https://github.com/restic/restic/issues/4179
Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I'm done! This pull request is ready for review.

wplapper committed:
issue 4179

Enhancement: create ability to sort output of restic ls -l by
name, size, atime, ctime, mtime, time(=mtime), X(=extension), extension

https://github.com/restic/restic/issues/4179

wplapper committed:
ls: create ability to sort output of restic ls -l

ls: create ability to sort output of restic ls -l by one of the following list
name, size, atime, ctime, mtime, time(=mtime), extension.
Reverse sorting has been implemtented as well.

MichaelEischer commented on cmd/restic/cmd_ls.go:
```diff
@@ -72,6 +81,8 @@ func init() {
 	flags.BoolVar(&lsOptions.HumanReadable, "human-readable", false, "print sizes in human readable format")
 	flags.BoolVar(&lsOptions.Ncdu, "ncdu", false, "output NCDU export format (pipe into 'ncdu -f -')")
+	flags.StringVarP(&lsOptions.Sort, "sort", "s", "name", "sort output by [name|size|time(=mtime)|atime|ctime|mtime|X(=extension)|extension]")
+	flags.BoolVarP(&lsOptions.Reverse, "reverse", "R", false, "reverse sorted output")
```
`reverse` only works in combination with `--sort`. Return an error if `--sort` wasn't specified.

wplapper replied:
fixed

MichaelEischer commented on cmd/restic/cmd_ls.go:
```diff
@@ -348,6 +359,15 @@ func runLs(ctx context.Context, opts LsOptions, gopts GlobalOptions, args []stri
+	}
+
+	collector := []ToSort{}
+	outputSort := true
```
The data format of restic guarantees that the output is implicitly sorted by name. There's no need to store all output in memory in that case.

wplapper replied:
check for name removed

MichaelEischer commented on cmd/restic/cmd_ls.go:
```diff
@@ -72,6 +81,8 @@ func init() {
 	flags.BoolVar(&lsOptions.Recursive, "recursive", false, "include files in subfolders of the listed directories")
 	flags.BoolVar(&lsOptions.HumanReadable, "human-readable", false, "print sizes in human readable format")
 	flags.BoolVar(&lsOptions.Ncdu, "ncdu", false, "output NCDU export format (pipe into 'ncdu -f -')")
+	flags.StringVarP(&lsOptions.Sort, "sort", "s", "name", "sort output by [name|size|time(=mtime)|atime|ctime|mtime|X(=extension)|extension]")
```
When invalid values are passed to `--sort`, the `ls` command must fail with an error. Take a look at `CompressionMode` in `internal/repository/repository.go` how a clean solution would look like.

MichaelEischer replied:
The general idea with parameter validation is that unsupported/invalid parameters must result in an error. That will allow us to support new values in the future without breaking existing behavior.

As a random example: For example `--sort mode` would right now sort by name as it isn't supported. But when we for some reason want to support `mode` in the future, then that would change the behavior for everyone who's for some reason already using `--sort mode`. By just returning an error when an invalid value was specified, we can ensure that nobody uses those invalid values and therefore are able to redefine the without problems in the future.

wplapper replied:
implemented

MichaelEischer commented on cmd/restic/cmd_ls.go:
```diff
@@ -444,5 +469,78 @@ func runLs(ctx context.Context, opts LsOptions, gopts GlobalOptions, args []stri
+				node      *restic.Node
+			}
+			collector_ext := make([]ToSortExt, len(collector))
+			for ix, item := range collector {
```
Is there a reason for using `ix` instead of the standard `i`?

wplapper replied:
code not longer there.

MichaelEischer commented on cmd/restic/cmd_ls.go:
```diff
@@ -444,5 +469,78 @@ func runLs(ctx context.Context, opts LsOptions, gopts GlobalOptions, args []stri
+	if outputSort {
+		if opts.Sort == "size" {
+			slices.SortStableFunc(collector, func(a, b ToSort) int {
+				return cmp.Or(
```
cmp.Or is not yet available in Go 1.21 (the current minimum Go version supported by restic). But we could just as well bump the minimum required go version to 1.22.

wplapper replied:
I prefer to be able to use SortStableFunc over Sort, since it is generic and also allows to use cmp.Or().

MichaelEischer commented on cmd/restic/cmd_ls.go:
```diff
@@ -444,5 +469,78 @@ func runLs(ctx context.Context, opts LsOptions, gopts GlobalOptions, args []stri
+			})
+		} else if opts.Sort == "X" || opts.Sort == "extension" {
+			// prepare sort by extension
+			type ToSortExt struct {
```
Using a `map[string]string` that maps the nodepath to the file extension would allow getting rid of these back and forth copies.

wplapper replied:
done. Got rid of the copy-in copy-out.

MichaelEischer commented on cmd/restic/cmd_ls.go:
```diff
@@ -444,5 +469,78 @@ func runLs(ctx context.Context, opts LsOptions, gopts GlobalOptions, args []stri
 		return err
 	}
 
+	if outputSort {
```
Please extract the sorted output code into a helper function. `runLs` is already far, far too long.

wplapper replied:
done

MichaelEischer commented on cmd/restic/cmd_ls.go:
```diff
@@ -444,5 +469,78 @@ func runLs(ctx context.Context, opts LsOptions, gopts GlobalOptions, args []stri
 	}
 
+	if outputSort {
+		if opts.Sort == "size" {
```
Use `switch opts.Sort {` here.

wplapper replied:
yep, done

MichaelEischer commented on cmd/restic/cmd_ls.go:
```diff
@@ -72,6 +81,8 @@ func init() {
 	flags.BoolVar(&lsOptions.Recursive, "recursive", false, "include files in subfolders of the listed directories")
 	flags.BoolVar(&lsOptions.HumanReadable, "human-readable", false, "print sizes in human readable format")
 	flags.BoolVar(&lsOptions.Ncdu, "ncdu", false, "output NCDU export format (pipe into 'ncdu -f -')")
+	flags.StringVarP(&lsOptions.Sort, "sort", "s", "name", "sort output by [name|size|time(=mtime)|atime|ctime|mtime|X(=extension)|extension]")
```
Where does `X` originate from? The other values largely match those offered by the unix `ls` command (`none, size, time, version, extension, width`). But `X` isn't part of that list.

wplapper replied:
X removed

MichaelEischer commented on cmd/restic/cmd_ls.go:
```diff
@@ -72,6 +81,8 @@ func init() {
 	flags.BoolVar(&lsOptions.HumanReadable, "human-readable", false, "print sizes in human readable format")
 	flags.BoolVar(&lsOptions.Ncdu, "ncdu", false, "output NCDU export format (pipe into 'ncdu -f -')")
+	flags.StringVarP(&lsOptions.Sort, "sort", "s", "name", "sort output by [name|size|time(=mtime)|atime|ctime|mtime|X(=extension)|extension]")
+	flags.BoolVarP(&lsOptions.Reverse, "reverse", "R", false, "reverse sorted output")
```
`-R` means `--recursive` for the `ls` command. So I'm wondering whether it makes sense to have a `-R` with a rather different meaning here.

wplapper replied:
removed short option -R

MichaelEischer commented on cmd/restic/cmd_ls.go:
```diff
@@ -36,6 +39,10 @@ will allow traversing into matching directories' subfolders.
 a path separator); paths use the forward slash '/' as separator.
 
+File listings can be sorted by specifying --sort followed by one of the
+sort specifiers '[name|size|time|atime|ctime|mtime]'.
```
`extension` is missing from the description.

wplapper replied:
extension has been added.

MichaelEischer reviewed (changes requested):
Thanks for working on this feature and the other PRs (I'll slowly dig through those).

I've left a few comments below. Please also add a basic test. You can look at `TestRunLsNcdu` for inspiration.

wplapper commented on changelog/unreleased/issue-4179:
```diff
@@ -0,0 +1,6 @@
+ls: create ability to sort output of restic ls -l by one of the following list
```
done

wplapper replied:
4 basic tests added. I had to expand `backup-data` to get some useful filenames with extensions. Ran all tests with the modified `backup-data` structure.

wplapper commented on cmd/restic/cmd_ls.go:
```diff
@@ -72,6 +81,8 @@ func init() {
 	flags.BoolVar(&lsOptions.Recursive, "recursive", false, "include files in subfolders of the listed directories")
 	flags.BoolVar(&lsOptions.HumanReadable, "human-readable", false, "print sizes in human readable format")
 	flags.BoolVar(&lsOptions.Ncdu, "ncdu", false, "output NCDU export format (pipe into 'ncdu -f -')")
+	flags.StringVarP(&lsOptions.Sort, "sort", "s", "name", "sort output by [name|size|time(=mtime)|atime|ctime|mtime|X(=extension)|extension]")
```
done

wplapper commented:
Regards testing: you have touched a sore point. Is there anywhere a document or a set of documents which describes the restic testing machinery? I am looking at TestRunLsNcdu and can see that it runs a backup to repo `env.testdata+"/0"`. from ".". What is `env.testdata+"/0"` and where does "." point to? Eventually the test function runs a couple of `runLs(context.TODO(), opts, gopts, args)` from the backup just generated.

I would like to create a test scenario where I backup 2 or 3 files with different names, extension and sizes. Then I can sort by name, time, size and extension and compare against the expected output.

Any pointers? Thanks!

wplapper committed:
restic ls - improve sorting enhancement code

Implemented all the changes that Michael Escher recommended.

As soon as I understand the workings of the test machinery, I will add testing code.

wplapper committed:
restic ls: add 4 integration tests for ls --sort

Extended standard backup-data by 3 files with extension names.

MichaelEischer reviewed (changes requested):
I have a few more nits. Can you add a short note about the new sort mode to `doc/045_working_with_repos.rst`?

I'll refactor the code a bit after merging, but it's harder to explain what I have in mind than making those changes.

MichaelEischer commented:
https://github.com/restic/restic/pull/5232 will bump the minimum go version to 1.22, this should fix the failing test in Go 1.21.

wplapper commented on changelog/unreleased/issue-4179:
```diff
@@ -0,0 +1,11 @@
+Enhancement: add capability sort output by one of given options
```
done.

wplapper commented on cmd/restic/cmd_ls.go:
```diff
@@ -444,5 +482,100 @@ func runLs(ctx context.Context, opts LsOptions, gopts GlobalOptions, args []stri
 	return printer.Close()
 }
+
+func PrintSortedOutput(printer lsPrinter, opts LsOptions, sortMode SortMode, collector []ToSortOutput) {
```
done.

wplapper replied:
done.

wplapper commented on cmd/restic/cmd_ls.go:
```diff
@@ -277,13 +288,33 @@ func (p *textLsPrinter) Close() error {
 }
 
+// for ls -l output sorting
+type ToSortOutput struct {
```
done.

wplapper replied:
done

wplapper commented on cmd/restic/cmd_ls.go:
```diff
@@ -277,13 +288,33 @@ func (p *textLsPrinter) Close() error {
+	}
+
+	sortMode := SortModeName
+	if opts.Sort != "" {
```
removed the if cause.

wplapper commented on cmd/restic/cmd_ls.go:
```diff
@@ -36,6 +39,10 @@ will allow traversing into matching directories' subfolders.
 a path separator); paths use the forward slash '/' as separator.
 
+File listings can be sorted by specifying --sort followed by one of the
+sort specifiers '(name|size|atime|ctime|mtime=time|extension)'.
```
I hope I changed it to be the same.

wplapper committed:
restic ls --sort

Incorporated the changes recommended by Michael.

wplapper committed:
restic ls documentation

Add a short node of hoe to use the --sort and --reverse options.

wplapper committed:
restic ls option --sort

Corrected lint errors.

MichaelEischer committed:
simplify sort mode handling

MichaelEischer committed:
fix linter errors

MichaelEischer committed:
tweak docs and remove user name

MichaelEischer committed:
drop username from metadata

MichaelEischer commented:
I've rebased the PR (to have Go 1.22 as minimum version), and made a few cleanups.

MichaelEischer commented:
Thanks for keeping up with all the review comments :-) . I've opened https://github.com/restic/restic/pull/5235 in case you're interested in the refactorings I had in mind.
